src/persist-txn/src/txn_read.rs

```diff
@@ -328,23 +393,88 @@ enum TxnsReadCmd<T> {
         tx: oneshot::Sender<DataListenNext<T>>,
     },
     Wait {
+        id: Uuid,
         ts: WaitTs<T>,
         tx: oneshot::Sender<()>,
     },
+    CancelWait {
+        id: Uuid,
+    },
 }
 
-#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[derive(Debug, PartialEq, Eq, Clone)]
 enum WaitTs<T> {
     GreaterEqual(T),
     GreaterThan(T),
 }
 
+// Specially made for keeping `WaitTs` in a `BTreeSet` and peeling them off in
+// the order in which they would be retired.
+//
+// MUSING: Adding more of this special-case logic around > and >= is maybe a
+// sign that the abstraction could be improved. Do we really need `update_ge`
+// and `update_gt`? Haven't thought about it too long because it's getting late.
+impl<T: Ord> Ord for WaitTs<T> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        let self_ts = match self {
+            WaitTs::GreaterEqual(ts) => ts,
+            WaitTs::GreaterThan(ts) => ts,
+        };
+        let other_ts = match other {
+            WaitTs::GreaterEqual(ts) => ts,
+            WaitTs::GreaterThan(ts) => ts,
+        };
+
+        if self_ts < other_ts {
+            Ordering::Less
+        } else if *self_ts > *other_ts {
+            Ordering::Greater
+        } else if matches!(self, WaitTs::GreaterEqual(_)) && matches!(other, WaitTs::GreaterThan(_))
+        {
+            // WIP: Are these two correct, for peeling of waits from
+            // `pending_wait_set`?
+            Ordering::Less
+        } else if matches!(self, WaitTs::GreaterThan(_)) && matches!(other, WaitTs::GreaterEqual(_))
+        {
+            Ordering::Greater
+        } else {
+            Ordering::Equal
+        }
+    }
+}
+
+impl<T: Ord> PartialOrd for WaitTs<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<T: Timestamp + Lattice + Codec64> WaitTs<T> {
+    /// Returns `true` iff (sic) this [WaitTs] is ready.
+    fn is_ready(&self, frontier: &T) -> bool {
+        match &self {
+            WaitTs::GreaterEqual(ts) => {
+                if frontier >= ts {
+                    return true;
+                }
+            }
+            WaitTs::GreaterThan(ts) => {
+                if frontier > ts {
+                    return true;
+                }
+            }
+        };
+
+        false
+    }
+}
+
 #[derive(Debug)]
 struct TxnsReadTask<T: Timestamp + Lattice + Codec64> {
     rx: mpsc::Receiver<TxnsReadCmd<T>>,
     cache: TxnsCacheState<T>,
-    pending_waits: Vec<PendingWait<T>>,
-    frontier: T,
+    pending_waits_set: BTreeSet<(WaitTs<T>, Uuid)>,
```

Comment:
`_set` and `_map` probably don't add a lot of value here. How about `_by_ts` and `_by_id` respectively?

Reply:
:+1: